### PR TITLE
Document gpio for f373 and f3x8

### PR DESCRIPTION
**Pull request by Sh3Rm4n**
Some GPIO peripherals do not have a `LCKR` register, so `gpio_v2` can not be
applied.

~`GPIO[AH]` and `GPIO[BD]` could be merged. For now, I've considered not to, as
they are split the same in the corresponding `svd` file. I can merge them,
if that's desired.~

See RM0313 8.4.8 and RM0316 11.4.8

### CODE REVIEW

**Comment by Sh3Rm4n** [comment]
Ok, this case is a bit confusing:

> x= A, B and D in STM32F303xB/C and STM32F358xC devices, x= A, B, C, D and F in
> STM32F303x6/8 and STM32F328x8 devices and x = A, B, C, D, E, F, G, H in
> STM32F303xD/E devices.

A `GPIOC` `LCKR` file should be available for the `STM32F328x8` but not for the
`STM32F358xC`. The `stm32f3x8.svd` file does not contain a LCKR description at all.

**Comment by aurelj** [comment]
> Ok, this case is a bit confusing:
> 
> > x= A, B and D in STM32F303xB/C and STM32F358xC devices, x= A, B, C, D and F in
> > STM32F303x6/8 and STM32F328x8 devices and x = A, B, C, D, E, F, G, H in
> > STM32F303xD/E devices.
> 
> A `GPIOC` `LCKR` file should be available for the `STM32F328x8` but not for the
> `STM32F358xC`. The `stm32f3x8.svd` file does not contain a LCKR description at all.
> 
> And the `STM32F398` is not mentioned at all. Is there a way to resolve this with patched svd's?

Yes, `GPIOC LCKR` should be available for the `STM32F328`.
To me the most sensible solution is to add LCKR register to GPIOC in stm32f3x8.yaml.
Then you can use gpio_v2.yaml for stm32f3x8.

And to document GPIO for stm32f373 without too much duplication, you could do something like this:

in gpio_v2.yaml:
```
_include:
 - gpio_v2_common.yaml

"GPIO*":
  LCKR:
    "LCK[0123456789]*":
      Unlocked: [0, "Port configuration not locked"]
      Locked: [1, "Port configuration locked"]
    "LCKK":
      NotActive: [0, "Port configuration lock key not active"]
      Active: [1, "Port configuration lock key active"]
```

add a gpio_v2_f373.yaml:
```
_include:
 - gpio_v2_common.yaml

"GPIO[ABD]":
  LCKR:
    "LCK[0123456789]*":
      Unlocked: [0, "Port configuration not locked"]
      Locked: [1, "Port configuration locked"]
    "LCKK":
      NotActive: [0, "Port configuration lock key not active"]
      Active: [1, "Port configuration lock key active"]
```

and create a gpio_v2_common.yaml, with all the current content of gpio_v2.yaml except for the LCKR things

**Comment by Sh3Rm4n** [comment]
Thanks for the help :)

I think, creatinga `gpio_v2_common` is the way to go. I will include your
suggestions.

> Yes, GPIOC LCKR should be available for the STM32F328.
> To me the most sensible solution is to add LCKR register to GPIOC in
> stm32f3x8.yaml.
> Then you can use gpio_v2.yaml for stm32f3x8.

That's what I would like to do as well. The `stm32f3x8.svd` however does not include a LCKR description for the GPIOC periphery.
Ignoring this, it would fail at:

```console
python3 scripts/svdpatch.py devices/stm32f3x8.yaml
Traceback (most recent call last):
  File "scripts/svdpatch.py", line 893, in <module>
    main()
  File "scripts/svdpatch.py", line 886, in main
    process_device(svd, root)
  File "scripts/svdpatch.py", line 864, in process_device
    d.process_peripheral(periphspec, device[periphspec],
  File "scripts/svdpatch.py", line 406, in process_peripheral
    p.process_register(rspec, register, update_fields)
  File "scripts/svdpatch.py", line 669, in process_register
    raise MissingRegisterError("Could not find {}:{}"
__main__.MissingRegisterError: Could not find GPIOC:LCKR
make: *** [Makefile:39: svd/stm32f3x8.svd.patched] Error 1
```

I've two options now:

1. Do not include GPIOC LCKR for the stm32f328
2. Or patch the svd, so GPIOC LCKR get's added?

**Comment by aurelj** [comment]
> I've two options now:
> 
>  1. Do not include GPIOC LCKR for the stm32f328
>  2. Or patch the svd, so GPIOC LCKR get's added?

I was actually suggesting option 2.

To achieve this, you should add something like this in stm32f3x8.yaml:
```
GPIOC:
  _add:
    LCKR:
      description: GPIO port configuration lock register
      addressOffset: 0x1C
      access: read-write
      fields:
        LCKK:
          description: Lok Key
          bitOffset: 16
          bitWidth: 1
        LCK15:
          description: Port x lock bit y (y=0..15)
          bitOffset: 15
          bitWidth: 1
```
and so on...

BTW: you forgot to add the `gpio_v2_common.yaml` and `gpio_v2_f373.yaml` file in your last commit...

**Comment by Sh3Rm4n** [comment]
Okay, nice. I did apply option 2. However, I wonder what happens, lckr is accesssed when using e.g. the stm32f398 or similar, which do not have the LCKR register.

File: devices/stm32f3x8.yaml [hunk 1]
```diff
@@ -4,6 +4,83 @@ _copy:
...
 
+GPIOC:
+  _add:
+    LCKR:
+      description: GPIO port configuration lock register
+      addressOffset: 0x1C
+      access: read-write
+      fields:
+        LCKK:
+          description: Lok Key
```
Comment (adamgreig) [inline devices/stm32f3x8.yaml]:
```suggestion
          description: Lock Key
```

**Comment by Sh3Rm4n** [comment]
Rebased and fixed the typo. Thank you for the review @adamgreig :)

**Pull request merged**